src/Engine/ProtoCore/RuntimeMemory.cs

```diff
@@ -294,32 +294,101 @@ public int CurrentConstructBlockId
                 }
             }
 
+
+            private int fp;
+            private StackFrame stackFrame;
+
             /// <summary>
             /// Current stack frame.
             /// </summary>
             public StackFrame CurrentStackFrame
             {
                 get
                 {
-                    var fp = FramePointer;
-
                     // Note the first stack frame starts after the space for global 
                     // variables, so here we need to check the frame pointer is 
                     // large enought to contain a stack frame and all global variables
-                    if (fp - StackFrame.StackFrameSize >= startFramePointer)
+                    if (FramePointer - StackFrame.StackFrameSize < startFramePointer)
                     {
+                        fp = 0;
+                        stackFrame = null;
+                        return null;
+                    }
+                       
+                    if (fp != FramePointer || stackFrame != null)
+                    {
+                        fp = FramePointer;
+
                         var frame = new StackValue[StackFrame.StackFrameSize];
-                        for (int sourceIndex = fp - 1, destIndex = 0; sourceIndex >= fp - StackFrame.StackFrameSize; sourceIndex--, destIndex++)
+                        for (int sourceIndex = fp - 1, destIndex = 0;
+                            sourceIndex >= fp - StackFrame.StackFrameSize;
+                            sourceIndex--, destIndex++)
                         {
                             frame[destIndex] = Stack[sourceIndex];
                         }
-                        var stackFrame = new StackFrame(frame);
+
+                        stackFrame = new StackFrame(frame);
                         return stackFrame;
                     }
-                    else
+
+                    return stackFrame;
+                }
+            }
+
+            public bool IsCurrentStackFrameNull
+            {
+                get
+                {
+                    if (FramePointer - StackFrame.StackFrameSize >= startFramePointer)
                     {
-                        return null;
+                        return false;
                     }
+
+                    return true;
+                }
```

Comment:
Why can't we simply say:
```
get { stackFrame == null; }
```

Reply:
I don't want to initialize the `stackFrame` object unless you actually need a `StackFrame` object.  That is the big time penalty here.  All of these essentially copy the implementation from StackFrame here to ovoid the getter for the CurrentStackFrame.  On my sample graph it is called 21million times.